### DISCUSSION TITLE
pdf.unbreakable() seems a bit broken

### DISCUSSION THREAD

**Discussion by macsourcerer** [py-pdf/fpdf2#557] [Q&A]
Hey,
before I file an issue I wonder if anyone else mentioned it.
if a
`with pdf_file.unbreakable() as pdf:`
causes a page break, the footer of the previous page and the header of the new page won't be added.
Also
`with pdf.local_context(fill_opacity=0.3):`
will not work on the newly created page.

Any help is appreciated.

**Comment by gmischler** [comment]
>  I wonder if anyone else mentioned it.

That's what the search function at the top of the page is there for.

If you file a bug as an issue, you will be presented with a template. Please make sure to provide all the information it asks for.
With a vague description like this one here it can be very hard to even just reproduce a possible problem, let alone to fix it.

**Reply by Lucas-C** [reply]
I agree with @gmischler : some code would be very welcome to be able to reproduce the problem you faced 😊

**Reply by macsourcerer** [reply]
I have a 800+ lines test file I can post here if its ok. Or should I file an issue?

**Reply by Lucas-C** [reply]
> I have a 800+ lines test file I can post here if its ok. Or should I file an issue?

Ideally, the best would be to post a <ins>**[minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example)**</ins>, which would mean to take the time to reduce your 800+ lines of code to something smaller

Then of course feel free to fill a bug report issue 😀

**Comment by macsourcerer** [comment]
ok. after some testing and debugging I found something.
fist, the header and footer thingy was my fault. I forgot to set the font color. My bad.
But unbreakable is really a bit broken.
I used your example from the documentation and added some code:

```python
from fpdf import FPDF

pdf = FPDF()
pdf.add_page()
pdf.set_font("Times", size=16)

line_height = pdf.font_size * 2
col_width = pdf.epw / 4

data = (
    ("First name", "Last name", "Age", "City"),
    ("Jules", "Smith", "34", "San Juan"),
    ("Mary", "Ramos", "45", "Orlando"),
    ("Carlson", "Banks", "19", "Los Angeles"),
    ("Lucas", "Cimon", "31", "Saint-Mahturin-sur-Loire"),
)

pdf.set_fill_color(255, 0, 0)

for i in range(5):
    with pdf.unbreakable() as doc:
        for row in data:
            with doc.local_context(fill_opacity=0.3):
                doc.cell(col_width, line_height, f"doc.get_y(): {int(doc.get_y())}", border=1, fill=True)
            for datum in row:
                doc.cell(col_width, line_height, f"{datum} ({i})", border=1)
            doc.ln(line_height)
    pdf.ln(line_height * 2)

pdf.output("unbreakable_tables.pdf")
```

1.
`with doc.local_context(fill_opacity=0.3):`
did not work after a page break.

2.
`doc.get_y()`
did not return the right y pos after a page break.
[unbreakable_tables.pdf](https://github.com/PyFPDF/fpdf2/files/9646355/unbreakable_tables.pdf)

**Reply by gmischler** [reply]
> `with doc.local_context(fill_opacity=0.3):` did not work after a page break.
> 
> `doc.get_y()` did not return the right y pos after a page break. [unbreakable_tables.pdf](https://github.com/PyFPDF/fpdf2/files/9646355/unbreakable_tables.pdf)

While those are technically two seperate issues, they do have the same root cause.
`.unbreakable()` is implemented through `FPDFRecorder()`, which again is kind of a metaprogramming hack to record all method calls within its managed context, and replay them on exit, either on the same page or the next. This approach worked reasonably well when fpdf2 was much simpler than it is today. It may or may not survive further development.

Within this framework, I don't think that your second issue can be fixed, or even needs to be fixed. Your code converts internal state of the library (`doc.y`) into text. The values that you see are the values during its execution, and are exactly correct at that moment. Just because the resulting text later gets moved to another page doesn't make them wrong.

The failure of `fill_opacity` is more complex in nature. It was added in the largely independent drawing.py module, and currently doesn't get stored on fpdf2's stack of local graphics context values where `FPDFRecorder()` would have access to it. I'm not sure how easily that can be fixed, but it would probably be a good idea to try.

**Reply by macsourcerer** [reply]
Thanks for the reply.
The `fill_opacity` issue is not important to me because we already get rid of it in our code.
The `doc.y` issue is really bad. It makes drawing lines or any other form at an relative y position impossible 
if there is an automatic page brake before.

**Reply by gmischler** [reply]
The auto-break functionality is primarily designed for text. If it works for some graphical content as well, then that's rather by accident...

But the solution for your use case is relatively straightforward.
Just determine yourself whether your stuff will fit on the remaining part of the page, and do a manual page break otherwise.

**Reply by Lucas-C** [reply]
I opened #562 to forbid the use of `get_y()` & `local_context()` inside `unbreakable()`, at least until we find a better solution.
What do you think @macsourcerer & @gmischler? Could you please add your comments on this PR?
